### PR TITLE
Add admin functionality for kafka rest

### PR DESCRIPTION
**Pull request by gabriel-tincu**
- Publish for topic / partition (even if it's already scope creep)
- Partitions and partitions details
- Broker info
- Topic list and details
- Stubs for consume and produce routes
- Unit tests for basic happy / sad flow
- Admin client addons to make it all work

### CODE REVIEW

File: karapace/kafka_rest_apis.py [hunk 1]
```diff
@@ -0,0 +1,680 @@
...
+                try:
+                    await self.validate_schema_info(data, prefix, content_type, topic)
+                except InvalidMessageSchema as e:
+                    self.r(body={"error_code": 42205, "message": str(e)}, content_type=content_type, status=422)
+
+    def produce_message(self, *, topic: str, key: bytes, value: bytes, partition: int = None) -> dict:
+        try:
+            with self.producer_lock:
+                f = self.producer.send(topic, key=key, value=value, partition=partition)
+                self.producer.flush()
```
Comment (vtainio) [inline karapace/kafka_rest_apis.py:570]: We should add a timeout

File: karapace/kafka_rest_apis.py [hunk 2]
```diff
@@ -0,0 +1,680 @@
...
+            )
+        except (SchemaRetrievalError, InvalidMessageSchema) as e:
+            self.r(body={"error_code": 42205, "message": str(e)}, content_type=content_type, status=422)
+        response = {
+            "key_schema_id": data.get("key_schema_id"),
+            "value_schema_id": data.get("value_schema_id"),
+            "offsets": []
+        }
+        for key, value, partition in prepared_records:
+            no_arg_produce = partial(self.produce_message, topic=topic, key=key, value=value, partition=partition)
```
Comment (vtainio) [inline karapace/kafka_rest_apis.py:355]: Maybe just run this normally?

File: karapace/kafka_rest_apis.py [hunk 3]
```diff
@@ -0,0 +1,678 @@
...
+class KafkaRest(KarapaceBase):
+    def __init__(self, config_path):
+        super().__init__(config_path)
+        self.serializer = SchemaRegistrySerializer(config_path=config_path)
+        self.deserializer = SchemaRegistryDeserializer(config_path=config_path)
+        self.kafka_timeout = 10
+        self.loop = asyncio.get_event_loop()
+        self._cluster_metadata = None
+        self._metadata_birth = None
+        # good as any
```
[the remark this answers is not in the document]
Comment (gabriel-tincu) [reply karapace/kafka_rest_apis.py:171]: just me picking my brain for a value and then just settling for one. I'll drop it as i realise it's confusing

File: karapace/kafka_rest_apis.py [hunk 4]
```diff
@@ -0,0 +1,678 @@
...
+            raise UnknownTopicOrPartitionError("Invalid values for offsets found")
+        return rv
+
+
+class KafkaRest(KarapaceBase):
+    def __init__(self, config_path):
+        super().__init__(config_path)
+        self.serializer = SchemaRegistrySerializer(config_path=config_path)
+        self.deserializer = SchemaRegistryDeserializer(config_path=config_path)
+        self.kafka_timeout = 10
```
[the remark this answers is not in the document]
Comment (gabriel-tincu) [reply karapace/kafka_rest_apis.py:167]: the `flush` method actually accepts a float parameter, so i believe this should remain as it is. It's also being used in the same manner in the registry apis code

File: karapace/kafka_rest_apis.py [hunk 5]
```diff
@@ -0,0 +1,678 @@
...
+    def get_offsets(self, topic, partition_id):
+        with self.admin_lock:
+            return self.admin_client.get_offsets(topic, partition_id)
+
+    def get_topic_config(self, topic):
+        with self.admin_lock:
+            return self.admin_client.get_topic_config(topic)
+
+    def cluster_metadata(self, topics=None):
+        if self._metadata_birth is not None and time.time() - self.metadata_max_age > self.metadata_max_age:
```
[the remark this answers is not in the document]
Comment (gabriel-tincu) [reply karapace/kafka_rest_apis.py:291]: fixed !

File: karapace/kafka_rest_apis.py [hunk 6]
```diff
@@ -0,0 +1,678 @@
...
+    def __init__(self, config_path):
+        super().__init__(config_path)
+        self.serializer = SchemaRegistrySerializer(config_path=config_path)
+        self.deserializer = SchemaRegistryDeserializer(config_path=config_path)
+        self.kafka_timeout = 10
+        self.loop = asyncio.get_event_loop()
+        self._cluster_metadata = None
+        self._metadata_birth = None
+        # good as any
+        self.metadata_max_age = 5
```
[the remark this answers is not in the document]
Comment (gabriel-tincu) [reply karapace/kafka_rest_apis.py]: done

**Pull request merged**